### PR TITLE
feat(inventory): add new option to not change asset states

### PR DESCRIPTION
**Pull request by stonebuzz**
Add new option to avoid the change of ```states_id``` (when is set manually before inventory)

image

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #number

### CODE REVIEW

File: src/Inventory/Asset/MainAsset.php [hunk 1]
```diff
@@ -645,7 +645,10 @@ public function rulepassed($items_id, $itemtype, $rules_id, $ports_id = 0)
         $entities_id = $this->entities_id;
         $val->is_dynamic = 1;
         $val->entities_id = $entities_id;
-        $val->states_id = $this->states_id_default ?? 0;
+
+        if ($this->states_id_default =! '-1') {
+            $val->states_id = $this->states_id_default ?? 0;
```
Comment (trasher) [inline src/Inventory/Asset/MainAsset.php:650]:
```suggestion
        if ($this->states_id_default ?? 0 =! '-1') {
            $val->states_id = $this->states_id_default ?? 0;
```

Comment (stonebuzz) [reply src/Inventory/Asset/MainAsset.php]: done

File: src/Inventory/Asset/MainAsset.php [hunk 2]
```diff
@@ -645,7 +645,10 @@ public function rulepassed($items_id, $itemtype, $rules_id, $ports_id = 0)
         $entities_id = $this->entities_id;
         $val->is_dynamic = 1;
         $val->entities_id = $entities_id;
-        $val->states_id = $this->states_id_default ?? 0;
+
+        if ($this->states_id_default ?? 0 != '-1') {
```
Comment (cedric-anne) [inline src/Inventory/Asset/MainAsset.php:649]: Maybe `0` value should be forced when we are in a new asset creation context, as this `0` value could probably be checked by other rules or plugins in other pieces of code.

File: src/Inventory/Asset/MainAsset.php [hunk 3]
```diff
@@ -645,7 +645,13 @@ public function rulepassed($items_id, $itemtype, $rules_id, $ports_id = 0)
...
         $val->is_dynamic = 1;
         $val->entities_id = $entities_id;
-        $val->states_id = $this->states_id_default ?? 0;
+
+        if ($items_id != 0 && $this->states_id_default != '-1') {
+            $val->states_id = $this->states_id_default ?? 0;
+        } elseif ($items_id == 0) {
+            //if create mode default states_id can't be '-1' put 0 if needed
+            $val->states_id = $this->states_id_default > 0 ? $this->states_id_default : 0;
+        }
```
Comment (cedric-anne) [inline src/Inventory/Asset/MainAsset.php:654]:
```suggestion
        $default_states_id = $this->states_id_default ?? 0;
        if ($items_id != 0 && $default_states_id != '-1') {
            $val->states_id = $default_states_id;
        } elseif ($items_id == 0) {
            //if create mode default states_id can't be '-1' put 0 if needed
            $val->states_id = $default_states_id > 0 ? $default_states_id : 0;
        }
```

File: src/Inventory/Conf.php [hunk 4]
```diff
@@ -450,12 +450,15 @@ public function showConfigForm()
         echo "</label>";
         echo "</td>";
         echo "<td>";
+
+        var_dump($config['states_id_default']);
```
Comment (orthagh) [inline src/Inventory/Conf.php:454]:
```suggestion
```

File: src/Inventory/Asset/MainAsset.php [hunk 5]
```diff
@@ -689,6 +695,8 @@ public function rulepassed($items_id, $itemtype, $rules_id, $ports_id = 0)
             $this->agent->fields['entities_id'] = $entities_id;
         }
 
+
+
```
Comment (orthagh) [inline src/Inventory/Asset/MainAsset.php:699]:
```suggestion
```

File: src/Inventory/Conf.php [hunk 6]
```diff
@@ -120,7 +120,7 @@ class Conf extends CommonGLPI
         'component_control'              => 1,
         'component_battery'              => 1,
         'component_simcard'              => 1,
-        'states_id_default'              => 0,
+        'states_id_default'              => -1,
```
[the remark this answers is not in the document]
Comment (stonebuzz) [reply src/Inventory/Conf.php:123]: people who add manually, then use the native inventory will not go to this option (but they won't have a problem)

People who only do native inventory, without touching this option will have the status empty (as if the default value was '----')

[the remark this answers is not in the document]
Comment (trasher) [reply src/Inventory/Conf.php:123]: That's the point: it changes the default value. Are we sure it won't be a problem that status will be kept in 10.0.4; but was set to default "0" value previousely

[the remark this answers is not in the document]
Comment (stonebuzz) [reply src/Inventory/Conf.php:123]: i see
agree with you

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Comment by TheRoby** [comment]
It's ok for me !
Thank you !
(but ... : the "do not change", is not translate ;)
And i think we have the same issue with network equipement (GLPI Inventory / config, network inventory)

But, again, very thank you !

**Comment by stonebuzz** [comment]
> It's ok for me ! Thank you ! (but ... : the "do not change", is not translate ;) And i think we have the same issue with network equipement (GLPI Inventory / config, network inventory)
> 
> But, again, very thank you !

Hi @TheRoby 

translation are done with Transifex =)

Best regards

**Pull request merged**